File: src/main/java/ru/citeck/ecos/notifications/domain/template/service/NotificationTemplateService.java
package ru.citeck.ecos.notifications.domain.template.service;

import lombok.Data;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.util.Assert;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.notifications.domain.template.converter.TemplateConverter;
import ru.citeck.ecos.notifications.domain.template.dto.MultiTemplateElementDto;
import ru.citeck.ecos.notifications.domain.template.dto.NotificationTemplateWithMeta;
import ru.citeck.ecos.notifications.domain.template.repo.NotificationTemplateEntity;
import ru.citeck.ecos.notifications.domain.template.repo.NotificationTemplateRepository;
import ru.citeck.ecos.records2.RecordConstants;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.predicate.PredicateUtils;
import ru.citeck.ecos.records2.predicate.model.Predicate;
import ru.citeck.ecos.records2.predicate.model.ValuePredicate;
import ru.citeck.ecos.records2.predicate.model.VoidPredicate;

import java.time.Instant;
import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

@Service
public class NotificationTemplateService {

    private final NotificationTemplateRepository templateRepository;
    private final TemplateConverter templateConverter;

    private Consumer<NotificationTemplateWithMeta> listener;


    public NotificationTemplateService(NotificationTemplateRepository templateRepository,
                                       TemplateConverter templateConverter) {
        this.templateRepository = templateRepository;
        this.templateConverter = templateConverter;
    }

    public void update(@NotNull NotificationTemplateWithMeta dto) {
        NotificationTemplateEntity template = templateRepository.save(templateConverter.dtoToEntity(dto));
        if (listener != null) {
            listener.accept(templateConverter.entityToDto(template));
        }
    }

    public void deleteById(String id) {
        if (StringUtils.isBlank(id)) {
            throw new IllegalArgumentException("Id parameter is mandatory for template deletion");
        }
        templateRepository.findOneByExtId(id).ifPresent(templateRepository::delete);
    }

    @NotNull
    public List<RecordRef> findTemplateRefsForTypes(@NotNull List<RecordRef> typeRefs) {

        if (typeRefs.isEmpty()) {
            return Collections.emptyList();
        }

        List<RecordRef> templates = new ArrayList<>();

        List<NotificationTemplateEntity> multiTemplates = templateRepository.findAllMultiTemplates();

        for (NotificationTemplateEntity template : multiTemplates) {
            NotificationTemplateWithMeta templateDto = templateConverter.entityToDto(template);
            if (templateDto.getMultiTemplateConfig() != null) {
                for (MultiTemplateElementDto configDto : templateDto.getMultiTemplateConfig()) {

                    if (RecordRef.isNotEmpty(configDto.getType())
                            && RecordRef.isNotEmpty(configDto.getTemplate())
                            && typeRefs.contains(configDto.getType())) {

                        templates.add(configDto.getTemplate());
                    }
                }
            }
        }

        return templates;
    }

    public Optional<NotificationTemplateWithMeta> findById(String id) {
        if (StringUtils.isBlank(id)) {
            return Optional.empty();
        }
        return templateRepository.findOneByExtId(id)
            .map(templateConverter::entityToDto);
    }

    public NotificationTemplateWithMeta save(NotificationTemplateWithMeta dto) {
        if (dto != null && !CollectionUtils.isEmpty(dto.getMultiTemplateConfig())) {
            List<MultiTemplateElementDto> checkedList = new ArrayList<>();
            for (MultiTemplateElementDto template : dto.getMultiTemplateConfig()) {
                if (template.getCondition() != null && !(template.getCondition() instanceof VoidPredicate)
                    || template.getTemplate() != null && RecordRef.isNotEmpty(template.getTemplate())
                    || template.getType() != null && RecordRef.isNotEmpty(template.getType())) {
                    checkedList.add(template);
                }
            }
            if (checkedList.size() != dto.getMultiTemplateConfig().size()) {
                dto.setMultiTemplateConfig(checkedList);
            }
        }
        if (dto != null && CollectionUtils.isEmpty(dto.getMultiTemplateConfig()))
            Assert.notEmpty(dto.getTemplateData(), "Template content must be defined");
        if (StringUtils.isBlank(dto.getId())) {
            dto.setId(UUID.randomUUID().toString());
        }
        NotificationTemplateEntity saved = templateRepository.save(templateConverter.dtoToEntity(dto));
        NotificationTemplateWithMeta resultDto = templateConverter.entityToDto(saved);
        if (listener != null) {
            listener.accept(resultDto);
        }
        return resultDto;
    }

    public List<NotificationTemplateWithMeta> getAll(int max, int skip, Predicate predicate, Sort sort) {
        if (sort == null) {
            sort = Sort.by(Sort.Direction.DESC, "id");
        }

        PageRequest page = PageRequest.of(skip / max, max, sort);

        return templateRepository.findAll(toSpec(predicate), page)
            .stream()
            .map(templateConverter::entityToDto)
            .collect(Collectors.toList());
    }

    private Specification<NotificationTemplateEntity> toSpec(Predicate predicate) {

        if (predicate instanceof ValuePredicate) {

            ValuePredicate valuePred = (ValuePredicate) predicate;

            ValuePredicate.Type type = valuePred.getType();
            Object value = valuePred.getValue();
            String attribute = valuePred.getAttribute();

            if (RecordConstants.ATT_MODIFIED.equals(attribute)
                && ValuePredicate.Type.GT.equals(type)) {

                Instant instant = Json.getMapper().convert(value, Instant.class);
                if (instant != null) {
                    return (root, query, builder) ->
                        builder.greaterThan(root.get("lastModifiedDate").as(Instant.class), instant);
                }
            }
        }

        PredicateDto predicateDto = PredicateUtils.convertToDto(predicate, PredicateDto.class);
        Specification<NotificationTemplateEntity> spec = null;

        if (org.apache.commons.lang3.StringUtils.isNotBlank(predicateDto.name)) {
            spec = (root, query, builder) ->
                builder.like(builder.lower(root.get("name")), "%" + predicateDto.name.toLowerCase() + "%");
        }
        if (org.apache.commons.lang3.StringUtils.isNotBlank(predicateDto.moduleId)) {
            Specification<NotificationTemplateEntity> idSpec = (root, query, builder) ->
                builder.like(builder.lower(root.get("extId")), "%" + predicateDto.moduleId.toLowerCase() + "%");
            spec = spec != null ? spec.or(idSpec) : idSpec;
        }

        return spec;
    }

    public long getCount(Predicate predicate) {
        Specification<NotificationTemplateEntity> spec = toSpec(predicate);
        return spec != null ? (int) templateRepository.count(spec) : getCount();
    }

    public long getCount() {
        return templateRepository.count();
    }

    public List<NotificationTemplateWithMeta> getAll(int max, int skip) {
        Sort sort = Sort.by(Sort.Direction.DESC, "id");
        PageRequest page = PageRequest.of(skip / max, max, sort);

        return templateRepository.findAll(page)
            .stream()
            .map(templateConverter::entityToDto)
            .collect(Collectors.toList());
    }

    public void addListener(Consumer<NotificationTemplateWithMeta> listener) {
        this.listener = listener;
    }

    @Data
    public static class PredicateDto {
        private String name;
        private String moduleId;
    }
}

File: src/main/java/ru/citeck/ecos/notifications/domain/template/service/NotificationTemplateService.kt
package ru.citeck.ecos.notifications.domain.template.service

import org.apache.commons.collections4.CollectionUtils
import org.apache.commons.lang3.StringUtils
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.util.Assert
import ru.citeck.ecos.notifications.domain.template.converter.TemplateConverter
import ru.citeck.ecos.notifications.domain.template.dto.MultiTemplateElementDto
import ru.citeck.ecos.notifications.domain.template.dto.NotificationTemplateWithMeta
import ru.citeck.ecos.notifications.domain.template.repo.NotificationTemplateEntity
import ru.citeck.ecos.notifications.domain.template.repo.NotificationTemplateRepository
import ru.citeck.ecos.records2.predicate.model.Predicate
import ru.citeck.ecos.records2.predicate.model.VoidPredicate
import ru.citeck.ecos.records3.record.dao.query.dto.query.SortBy
import ru.citeck.ecos.webapp.api.entity.EntityRef
import ru.citeck.ecos.webapp.lib.spring.hibernate.context.predicate.JpaSearchConverter
import ru.citeck.ecos.webapp.lib.spring.hibernate.context.predicate.JpaSearchConverterFactory
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors
import javax.annotation.PostConstruct

// TODO: https://citeck.atlassian.net/browse/ECOSCOM-4811
@Service("domainNotificationTemplateService")
class NotificationTemplateService(
    private val templateRepository: NotificationTemplateRepository,
    private val templateConverter: TemplateConverter,
    private val jpaSearchConverterFactory: JpaSearchConverterFactory
) {

    private lateinit var searchConv: JpaSearchConverter<NotificationTemplateEntity>

    private var listener: Consumer<NotificationTemplateWithMeta>? = null

    val count: Long
        get() = templateRepository.count()

    @PostConstruct
    fun init() {
        searchConv = jpaSearchConverterFactory.createConverter(NotificationTemplateEntity::class.java).build()
    }

    fun update(dto: NotificationTemplateWithMeta) {
        val template = templateRepository.save(
            templateConverter.dtoToEntity(dto)
        )
        if (listener != null) {
            listener!!.accept(templateConverter.entityToDto(template))
        }
    }

    fun deleteById(id: String?) {
        require(!StringUtils.isBlank(id)) { "Id parameter is mandatory for template deletion" }
        templateRepository.findOneByExtId(id)
            .ifPresent { entity: NotificationTemplateEntity -> templateRepository.delete(entity) }
    }

    fun findTemplateRefsForTypes(typeRefs: List<EntityRef>): List<EntityRef> {
        if (typeRefs.isEmpty()) {
            return emptyList()
        }

        val templates: MutableList<EntityRef> = ArrayList()

        templateRepository.findAllMultiTemplates().forEach { notificationTemplateEntity ->
            templateConverter.entityToDto(notificationTemplateEntity).multiTemplateConfig?.let { multiTemplates ->
                for ((template, type) in multiTemplates) {
                    if (EntityRef.isNotEmpty(type) &&
                        EntityRef.isNotEmpty(template) &&
                        typeRefs.contains(type)
                    ) {
                        template?.let { templates.add(it) }
                    }
                }
            }
        }

        return templates
    }

    fun findById(id: String): Optional<NotificationTemplateWithMeta> {
        return if (StringUtils.isBlank(id)) {
            Optional.empty()
        } else {
            templateRepository.findOneByExtId(id)
                .map { entity: NotificationTemplateEntity ->
                    templateConverter.entityToDto(
                        entity
                    )
                }
        }
    }

    fun save(dto: NotificationTemplateWithMeta): NotificationTemplateWithMeta {
        if (CollectionUtils.isNotEmpty(dto.multiTemplateConfig)) {
            val notEmptyMultiTemplateConfigs: MutableList<MultiTemplateElementDto> = ArrayList()
            for (template in dto.multiTemplateConfig!!) {
                if (template.condition != null && template.condition !is VoidPredicate ||
                    EntityRef.isNotEmpty(template.template) || EntityRef.isNotEmpty(template.type)
                ) {
                    notEmptyMultiTemplateConfigs.add(template)
                }
            }
            if (notEmptyMultiTemplateConfigs.size != dto.multiTemplateConfig!!.size) {
                dto.multiTemplateConfig = notEmptyMultiTemplateConfigs
            }
        }

        if (CollectionUtils.isEmpty(dto.multiTemplateConfig)) {
            Assert.notEmpty(dto.templateData, "Template content must be defined")
        }

        if (StringUtils.isBlank(dto.id)) {
            dto.id = UUID.randomUUID().toString()
        }

        val saved = templateRepository.save(
            templateConverter.dtoToEntity(
                dto
            )
        )
        val resultDto = templateConverter.entityToDto(saved)
        if (listener != null) {
            listener!!.accept(resultDto)
        }

        return resultDto
    }

    fun getAll(max: Int, skip: Int, predicate: Predicate, sort: List<SortBy>): List<NotificationTemplateWithMeta> {
        return searchConv.findAll(templateRepository, predicate, max, skip, sort)
            .map { entity: NotificationTemplateEntity ->
                templateConverter.entityToDto(
                    entity
                )
            }
            .toList()
    }

    fun getCount(predicate: Predicate): Long {
        return searchConv.getCount(templateRepository, predicate)
    }

    fun getAll(max: Int, skip: Int): List<NotificationTemplateWithMeta> {
        val sort = Sort.by(Sort.Direction.DESC, "id")
        val page = PageRequest.of(skip / max, max, sort)
        return templateRepository.findAll(page)
            .stream()
            .map { entity: NotificationTemplateEntity ->
                templateConverter.entityToDto(
                    entity
                )
            }
            .collect(Collectors.toList())
    }

    fun addListener(listener: Consumer<NotificationTemplateWithMeta>) {
        this.listener = listener
    }
}
